overlord/snapstate/snapstate.go

```diff
@@ -1110,6 +1120,62 @@ func RevertToRevision(st *state.State, name string, rev snap.Revision, flags Fla
 	return doInstall(st, &snapst, snapsup)
 }
 
+// TransitionCore transitions from an old core snap name to a new core
+// snap name. It is used for the ubuntu-core -> core transition (that
+// is not just a rename because the two snaps have different snapIDs)
+//
+// Note that this function makes some assumptions like:
+// - no aliases setup for both snaps
+// - no data needs to be copied
+// - all interfaces are absolutely identical on both new and old
+// Do not use this as a general way to transition from snap A to snap B.
+func TransitionCore(st *state.State, oldName, newName string) ([]*state.TaskSet, error) {
+	var oldSnapst, newSnapst SnapState
+	err := Get(st, oldName, &oldSnapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	if !oldSnapst.HasCurrent() {
+		return nil, fmt.Errorf("cannot transition snap %q: not installed", oldName)
+	}
+
+	// FIXME: what user ID to use for an automatic transition?
+	userID := 0
+	newInfo, err := snapInfo(st, newName, oldSnapst.Channel, snap.R(0), userID)
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME: deal with the case that "newName" is already installed
+	// FIXME2: add missing tests
+
+	// start by instaling the new snap
+	tsInst, err := doInstall(st, &newSnapst, &SnapSetup{
+		Channel:      oldSnapst.Channel,
+		DownloadInfo: &newInfo.DownloadInfo,
+		SideInfo:     &newInfo.SideInfo,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// then transition the interface connections over
+	transIf := st.NewTask("transition-connections-core-migration", fmt.Sprintf(i18n.G("Transition security profiles from %q to %q"), oldName, newName))
+	transIf.Set("old-name", oldName)
+	transIf.Set("new-name", newName)
+	transIf.WaitAll(tsInst)
+	tsTrans := state.NewTaskSet(transIf)
+
+	// then remove the old snap
```

Comment:
while we have both installed I'm not sure snapstate.CoreInfo will do the sanest thing, we need to teach it to pick a preferred one or something I suppose (not sure it does atm)